### PR TITLE
[Backport 2.x] MultiDataSource feature merge (#2334)

### PR DESCRIPTION
**Pull request by mpabba3003**
Signed-off-by: mpabba3003 <amazonmanideep@gmail.com>

### Description
[Backport 2.x] MultiDataSource feature merge (#2334)

Backport https://github.com/opensearch-project/OpenSearch-Dashboards/commit/890a4b649c0a4169a8683a7ca72c0fda55e886d0 from https://github.com/opensearch-project/OpenSearch-Dashboards/pull/2334

### CODE REVIEW

**Comment by ananzh** [comment]
Hey @mpabba3003 is this a duplicate backport PR of https://github.com/opensearch-project/OpenSearch-Dashboards/pull/2399? 

If so, pls close this one. Thx.

**Comment by seraphjiang** [comment]
> Hey @mpabba3003 is this a duplicate backport PR of #2399?
> 
> If so, pls close this one. Thx.

@ananzh we should close #2399 .
@mpabba3003  is from original contributors for the feature

**Comment by ananzh** [comment]
I see the previous backport PR (https://github.com/opensearch-project/OpenSearch-Dashboards/pull/2399) is closed. Please make sure to solve all the review comments from https://github.com/opensearch-project/OpenSearch-Dashboards/pull/2399 via following fix PR and merge to main first. Thx

File: src/plugins/data_source/config.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    wrappingKey: schema.arrayOf(schema.number(), {
+      minSize: WRAPPING_KEY_SIZE,
+      maxSize: WRAPPING_KEY_SIZE,
+      defaultValue: new Array(32).fill(0),
+    }),
+  }),
+  clientPool: schema.object({
+    size: schema.number({ defaultValue: 5 }),
+  }),
+  audit: schema.object({
```
Comment (kavilla) [inline src/plugins/data_source/config.ts:36]: nit: how come we want enable multiple log files? would there be a reason to just utilize existing auditing functionality?

File: src/plugins/index_pattern_management/public/components/create_index_pattern_wizard/components/step_index_pattern/step_index_pattern.tsx [hunk 2]
```diff
@@ -116,13 +126,16 @@ export class StepIndexPattern extends Component<StepIndexPatternProps, StepIndex
 
   ILLEGAL_CHARACTERS = [...indexPatterns.ILLEGAL_CHARACTERS];
 
+  dataSrouceEnabled: boolean;
```
Comment (kavilla) [inline src/plugins/index_pattern_management/public/components/create_index_pattern_wizard/components/step_index_pattern/step_index_pattern.tsx:129]: typo

**Review by kavilla — APPROVED** [review]
It's fine, there are follow-up requests related to UX/UI, bugs, testing, and improvements. Another improvement we can do is add your config to this file:

https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/bin/opensearch-dashboards-docker#L152

and then:
https://github.com/opensearch-project/opensearch-build/blob/main/docker/release/config/opensearch-dashboards/opensearch-dashboards-docker-entrypoint.sh#L156

File: src/plugins/data_source/config.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+import { schema, TypeOf } from '@osd/config-schema';
+import { fileAppenderSchema } from './audit_config';
+
+const KEY_NAME_MIN_LENGTH: number = 1;
+const KEY_NAME_MAX_LENGTH: number = 100;
+// Wrapping key size shoule be 32 bytes, as used in envelope encryption algorithms.
+const WRAPPING_KEY_SIZE: number = 32;
+
+export const configSchema = schema.object({
+  enabled: schema.boolean({ defaultValue: false }),
```
[the remark this answers is not in the document]
Comment (zhongnansu) [reply src/plugins/data_source/config.ts:15]: fixed in this PR #https://github.com/opensearch-project/OpenSearch-Dashboards/pull/2428

File: src/plugins/data_source/config.ts (hunk 1 above)
Comment (zhongnansu) [reply src/plugins/data_source/config.ts:36]: audit log is different with service log, for data source we'll have both. You can check #1986  for more discussion. cc @kristenTian

File: src/plugins/index_pattern_management/public/components/create_index_pattern_wizard/components/step_index_pattern/step_index_pattern.tsx (hunk 2 above)
Comment (zhongnansu) [reply src/plugins/index_pattern_management/public/components/create_index_pattern_wizard/components/step_index_pattern/step_index_pattern.tsx:129]: good catch

**Review by zengyan-amazon — APPROVED** [review]

**Pull request merged**